compiler-lib/src/sem.rs

```diff
@@ -1,6 +1,163 @@
+                        res.insert(
+                            (Rc::clone(&class.name), Rc::clone(&member.name), discr),
+                            &member,
+                        );
+                    }
+                }
+                Ok(())
+            }
+            x => {
+                return Err(format_err!(
+                    "unexpected ast node passed to first pass: {:?}",
+                    x
+                ))
+            }
+        }
+    }
+}
+
+type ScopeDefs<T> = HashMap<Symbol, T>;
+
+struct Scope<T> {
+    parent: Option<Rc<Scope<T>>>,
+    definitions: ScopeDefs<T>,
+}
+
+impl<T> Scope<T> {
+    pub fn root() -> Rc<Self> {
+        Rc::new(Scope {
+            parent: None,
+            definitions: ScopeDefs::new(),
+        })
+    }
+    pub fn enter_scope(parent: Rc<Self>) -> Self {
+        Scope {
+            parent: Some(parent),
+            definitions: ScopeDefs::new(),
+        }
+    }
+    pub fn leave(self) -> Option<Rc<Self>> {
+        self.parent
+    }
+    pub fn define(&mut self, sym: &Symbol, val: T) -> Result<(), ()> {
+        use std::collections::hash_map::Entry;
+        let key = Rc::clone(sym);
+        match self.definitions.entry(key) {
+            Entry::Occupied(o) => return Err(()),
+            Entry::Vacant(e) => e.insert(val),
+        };
+        Ok(())
+    }
+}
+
+struct SecondPass;
+
+// TODO proper enum, but Type is definitely necessary
+type VarDef<'a, 'f> = &'a Spanned<'f, ast::Type>;
+struct TyDef; // TODO
+struct MethDef; // TODO
+
+impl SecondPass {
+    pub fn visit<'a, 'f>(
+        ast: &'a ast::AST<'f>,
+        context: &context::Context<'_>,
+        _classes_and_members: &ClassesAndMembers<'a, 'f>,
+    ) -> Result<(), Error> {
+        // TODO java.lang.System.out scope
+        // TODO ontop of that, scope of this package
+        let mut vardefs: Rc<Scope<VarDef<'a, 'f>>> = Scope::root();
+        let mut tydefs: Rc<Scope<TyDef>> = Scope::root();
+        let mut methdefs: Rc<Scope<MethDef>> = Scope::root();
+        Self::do_visit(&NodeKind::from(ast), context, vardefs, tydefs, methdefs)
+    }
+
+    fn do_visit<'a, 'f>(
+        n: &NodeKind<'a, 'f>,
+        context: &context::Context<'_>,
+        vardefs: Rc<Scope<VarDef<'a, 'f>>>,
+        tydefs: Rc<Scope<TyDef>>,
+        methdefs: Rc<Scope<MethDef>>,
+    ) -> Result<(), Error> {
+        use self::NodeKind::*;
+        match n {
+            ClassMember(m) if m.kind.is_method() => {
+                // Assume types have been checked
+                // TODO enforce somewhere that main has String[] arg type
+                let params = m.kind.method_params().unwrap();
+                let mut vardefs = Scope::enter_scope(vardefs);
+                for p in &params.data {
+                    vardefs
+                        .define(&p.name, &p.ty)
+                        .or(Err(format_err!("re-definition of variable in same scope")))?;
```

Comment:
what's the problem here??? you can generate a normal derive(Fail) error here rendering the second definition using p.span.
Or do you want to attach 2 spans pointing at both definitions? We could implement multispans.

Reply:
>  we even need a Result if we just use Context to highlight sutff?

No. You just need anything that can be a string or has Display implemented. You could just do:
```
context.diagnostics.error(Spanned {
    span: p.span,
    data: SemanticError::RedefinitionError {
         kind: "variable", name: p.name.to_string()
    }
})
```
or: 
```
context.diagnostics.error(Spanned {
    span: p.span,
    data: "re-definition of variable in same scope"
})
```